### PR TITLE
Set up the project and add plugin modules with basic functionality

### PR DESCRIPTION
**Pull request by Aniket21mathur**
This Pr implements a basic working plugin, able to generate a `chpl` file. I have followed the general structure that other plugins have. The plugin files include -

- `protoc-gen-chpl.cc` - Calls the `PluginMain` function which handles the `CodeGeneratorRequest` and `CodeGeneratorResponse`, results in a generation of a `protoc-gen-chpl` binary file which is passed to `protoc` compiler.

- `generator.cc/hh` - It contains an inheritance class of `CodeGenerator` protobuf compiler class. Handles the generation of the output stream and calls the reflection class to generate the content of the output file.

- `reflection_class.cc/hh` - Initializes the content generation of the output file, will have basic classes to write descriptor and introductory code, make a call to the message class for each message.

- `message.cc/hh` - Supports classes to write a reflection of each message, make calls to the field reflection generators. 

- `field_base.cc/hh` - Set basic field properties that are common to all types of fields(required, optional, repeated, any, oneOf). Its base class is inherited by the various field type generator classes.

- `primitive_field.cc/hh` - Generate reflection of a message field, an inheritance of `field_base` class.

- `helpers.cc/hh` - Have helpers functions for the plugin, like converting to camel case naming convention.

To start with, these files currently have a very basic functionality implemented, just to support `int32` and `int64` types.

I have generated some [example](https://github.com/Aniket21mathur/ProtocolBuffers/tree/master/chapel/build) `chpl` files from these [source](https://github.com/Aniket21mathur/ProtocolBuffers/tree/master/chapel/src) files using the protoc compiler.

`protoc -I=src --chpl_out=build src/somefile.proto`

This pr also add `autotools`support  to build the project.

**I have not focused much on what should be the generated chapel code for `int` types, there is a little more discussion needed for that. My primary motive was to set up the architecture to let us begin with the code generation process**

### CODE REVIEW

**Comment by Aniket21mathur** [comment]
@mppf @lydia-duncan @astatide please have a look :)

**Comment by Aniket21mathur** [comment]
@lydia-duncan it would be great if you have a rough look at the file names, namespace, and function names for the plugin or might give overall feedback on the pr, whenever you get time :)

File: src/helpers.cc [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+#include <google/protobuf/stubs/strutil.h>
+
+#include <helpers.hh>
+#include <primitive_field.hh>
+#include <field_base.hh>
+
+namespace chapel {
+  
+  string StripDotProto(const std::string& proto_file) {
+    int lastindex = proto_file.find_last_of(".");
```
Comment (lydia-duncan) [inline src/helpers.cc:12]: We should be sure to verify that we don't get sent a file without an extension (or with an extension other than `.proto`)

File: src/helpers.hh [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+namespace chapel {
+  
+  using namespace std;
+  
+  using namespace google::protobuf;
+  using namespace google::protobuf::compiler;
+
+  class FieldGeneratorBase;
+  
+  string StripDotProto(const string& proto_file);
```
Comment (lydia-duncan) [inline src/helpers.hh]: This function doesn't seem to be used outside of a helper function that is not declared here.  Should its declaration also be internal to helpers.cc?

File: src/helpers.cc [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  
+  string GetFileNameBase(const FileDescriptor* descriptor) {
+      string proto_file = descriptor->name();
+      int lastslash = proto_file.find_last_of("/");
+      string base = proto_file.substr(lastslash + 1);
+      return UnderscoresToCamelCase(StripDotProto(base));
+  }
+
+  string GetOutputFile(const FileDescriptor* descriptor, string* error) {
+    string relative_filename = GetFileNameBase(descriptor);
```
Comment (lydia-duncan) [inline src/helpers.cc:66]: This seems to always result in the .chpl file being placed in the current directory?  I may not have followed the call chain back far enough, though

**Review by lydia-duncan — APPROVED** [review]
I noted a couple of edge cases to be aware of, and a slight nit pick.  Other than that, I think this is great and could be merged :)

File: src/helpers.cc (hunk 3 above)
Comment (Aniket21mathur) [reply src/helpers.cc:66]: No, the placement of the `.chpl` file is decided by the `GeneratorContext` object which deals with the output path passed to the `protoc` compiler as

`protoc --chpl_out=/output/dir/ a.proto`

This is an internal implementation in `google/protobuf/compiler/code_generator.cc`.

File: src/helpers.hh (hunk 2 above)
Comment (Aniket21mathur) [reply src/helpers.hh]: Yes, we can remove that. Thanks for pointing out :)

File: src/helpers.cc (hunk 1 above)
Comment (Aniket21mathur) [reply src/helpers.cc:12]: I added a check in 38323fd, please have a look

**Comment by Aniket21mathur** [comment]
@lydia-duncan thanks a lot for reviewing, I have made the suggested changes :)

**Review by lydia-duncan — APPROVED** [review]
Looks good :)

**Comment by Aniket21mathur** [comment]
Should I merge this pull request now?

**Comment by lydia-duncan** [comment]
Sure!

**Pull request merged**